src/strings/ops.c

```diff
@@ -619,71 +657,50 @@ MVMint64 MVM_string_equal_at_ignore_case(MVMThreadContext *tc, MVMString *haysta
-        const MVMCodepoint* h_result_cps;
-        h_g = MVM_string_get_grapheme_at_nocheck(tc, haystack, h_offset + i);
-        if (h_g >= 0 ) {
-            /* For codeponits we can get the case change directly */
-            h_fc_cps = MVM_unicode_get_case_change(tc, h_g, MVM_unicode_case_change_type_fold, &h_result_cps);
-        }
-        else {
-            /* Synthetics must use this function */
-            h_fc_cps = MVM_nfg_get_case_change(tc, h_g, MVM_unicode_case_change_type_fold, &h_result_cps);
-        }
-        /* If we get 0 for the number that means the cp doesn't change when casefolded */
-        if (h_fc_cps == 0) {
-            n_g = MVM_string_get_grapheme_at_nocheck(tc, needle_fc, i + n_offset);
-            if (h_g != n_g)
-                return 0;
-        }
-        else if (h_fc_cps >= 1) {
-            for (j = 0; j < h_fc_cps; j++) {
-                n_g = MVM_string_get_grapheme_at_nocheck(tc, needle_fc, i + n_offset);
-                h_g = h_result_cps[j];
-                if (h_g != n_g)
-                    return 0;
-                n_offset++;
-            }
-            n_offset--;
-        }
-    }
-    return 1;
+    return MVM_string_equal_at_ignore_case_INTERNAL_loop(tc, haystack, needle_fc, h_offset, h_graphs, n_graphs);
 }
 MVMint64 MVM_string_index_ignore_case(MVMThreadContext *tc, MVMString *haystack, MVMString *needle, MVMint64 start) {
+    /* foldcase version of needle */
+    MVMString *needle_fc;
+    MVMStringIndex n_fc_graphs;
+
     size_t index           = (size_t)start;
     MVMStringIndex hgraphs, ngraphs;
     MVMint64 return_val = -1;
     MVM_string_check_arg(tc, haystack, "index search target");
     MVM_string_check_arg(tc, needle, "index search term");
     hgraphs = MVM_string_graphs(tc, haystack);
     ngraphs = MVM_string_graphs(tc, needle);
-
     if (!ngraphs)
         return start <= hgraphs ? start : -1; /* the empty string is in any other string */
-
     if (!hgraphs)
         return -1;
-
     if (start < 0 || start >= hgraphs)
         return -1;
+    /* Codepoints can expand into up to THREE codepoints (as of Unicode 9.0). The next check
+     * checks if it is at all possible for the needle grapheme number to be higher
+     * than the haystack */
+    if (ngraphs > hgraphs * 3)
+        return -1;
 
-    if (ngraphs > hgraphs || ngraphs < 1)
+    if (ngraphs < 1)
         return -1;
 
+    needle_fc = MVM_string_fc(tc, needle);
+    n_fc_graphs = MVM_string_graphs(tc, needle_fc);
+
     MVMROOT(tc, haystack, {
-        MVMROOT(tc, needle, {
-            /* brute force for now. horrible, yes. halp. */
-            while (index <= hgraphs) {
-                if (MVM_string_equal_at_ignore_case(tc, haystack, needle, index)) {
-                    return_val = (MVMint64)index;
-                    break;
-                }
-                index++;
-            }
+        MVMROOT(tc, needle_fc, {
```

Comment:
Well, you are going to using `needle_fc` var again, but since `MVM_string_fc` may trigger a GC(which may moves address of objects), you need `MVMROOT` it(to get a new address of the object) , otherwise the address of `needle_fc` will be outdated, see https://github.com/samcv/MoarVM/blob/b9104537ce89bdee1d33905bb3de9e5a60457875/src/strings/ops.c#L654-L660 for example.